### PR TITLE
feat: stance evaluation, prompt regression, and OKF theme curation (#96, #97)

### PR DESCRIPTION
**Pull request by dceoy**
## Summary

Implements the remaining Phase 4 accountability machinery of roadmap #98 — the two currently-open linked issues #96 and #97 — while preserving every existing invariant (deterministic artifacts, hand-rolled validators, 100% branch coverage, dependency-light code, no vector DB/embeddings/RAG, and numeric market facts never authored by an LLM).

**This PR does not use `Closes` for either issue.** Both issues have acceptance-criteria milestones that depend on accumulated production data which does not exist yet: no qualitative artifacts have been committed to `main` (shadow mode has not produced them). Per the task brief, the machinery, schemas, fixtures, docs, and safe empty-state behavior are complete and the data-dependent residuals are recorded below rather than fabricated. Residual blockers are also posted on #96, #97, and #98.

## #97 — Evaluate stances against realized outcomes + prompt regression

- **`src/aims/performance.py`** joins per-instrument stances from committed `data/qualitative/` artifacts with realized forward returns. Forward returns are reconstructed deterministically by chaining each symbol's trailing `ret_1d` feature across analysis artifacts, keyed by the per-symbol bar date in `data_freshness` — **no live price fetches**. The chain self-checks against any later `ret_5d` feature; windows that fail are skipped (`broken_chain`), never scored with wrong numbers. Per-horizon (1d/5d/20d) supportive/neutral/conflicting hit rates, average returns, and confidence calibration. Writes the schema-validated `data/performance/<date>.json` artifact and the public `content/evaluation/_index.md` page. Reuses `history.py`/`backtest.py` conventions rather than a parallel design.
- **`src/aims/prompt_regression.py`** recomputes the validator and the #93 gates over a `(qualitative, analysis, evidence)` triple and asserts **structural and gate metrics** (validator cleanliness, market-narrative rendering, gate pass rate, citation coverage, stance-distribution sanity) — never exact prose. Results are recorded in `data/performance/prompt_regressions.json`, keyed by prompt SHA-256 + model ID. A CI test fails if the committed prompt/model has no recorded passing run, so a prompt/model change cannot merge without a recorded harness run. No automatic prompt tuning.
- **`src/aims/link_check.py`** samples recent evidence citation URLs and warns on link rot — **warning only**: always exits 0, writes no artifact, runs `continue-on-error` in CI.
- Hand-rolled validators (`validate_performance.py`, `validate_prompt_regression.py`) + JSON schema references. `notify_slack.py` gains a trailing AI-stance hit-rate line.
- Not an investable/executable track record: informational-association disclaimers are embedded in every artifact and rendered prominently.

## #96 — Curate durable themes into the OKF layer

- **`src/aims/okf_curation.py`** behind the `aims-okf-curator` skill (`curate_themes.py`): deterministic monthly pass clustering recurring theme titles, proposing promotions that recur on **≥3 distinct dates spanning ≥14 days** with dated artifact + evidence citations and a ready-to-edit concept skeleton, plus per-instrument stance streaks and **>60-day** retirement candidates. Output is a human-reviewed proposal only — it never writes to `okf/`, and OKF changes are never auto-merged. Cadence, promotion, and retirement criteria documented in the skill; `okf/logs/log.md` updated and shadow content regenerated via `tools/okf_hugo_adapter.py`.

## Workflow

`daily-market-analysis.yml` now evaluates stances, validates the artifact, regenerates the evaluation page, samples citation links (non-fatal), and commits `data/performance/` and `content/evaluation/` in the analysis PR. New `Evaluation` menu entry.

## Validation

Commands run locally from the repo root:

- `uv run ruff format .` — clean
- `uv run ruff check .` — All checks passed
- `uv run pyright .` — 0 errors
- `uv run pytest` — **926 passed, 100% branch coverage** (`--cov-fail-under=100` enforced; every new `src/aims` module at 100%)
- `uv run python tools/okf_hugo_adapter.py --src okf --dst content/knowledge --clean` then `--check` — passes; shadow content regenerated, not hand-edited
- `uvx yamllint`, `actionlint`, `uvx zizmor --offline` on `daily-market-analysis.yml` — no findings
- New/extended tests cover normal, empty, stale (broken-chain), invalid, and boundary cases; goldens added for the evaluation page (populated + empty) and the curation proposal (populated + empty)

Not run locally (binaries unavailable in this environment; both run in CI): `hugo --gc --minify` and `checkov`. The evaluation page's TOML front matter and the menu config were validated with `tomllib`; the OKF `--check` covers shadow-content correctness.

## Residual blockers (data-dependent, recorded not fabricated)

- **#96:** "at least one theme concept promoted through a reviewed PR" is not yet possible — zero qualitative artifacts on `main`. The first curation pass is a recorded empty "no promotions" pass (`okf/logs/log.md`). Machinery + criteria complete.
- **#97:** the committed `data/performance/2026-07-04.json` and `content/evaluation/_index.md` are the safe empty state (zero matured observations) and fill in as artifacts accumulate. A baseline prompt-regression entry is recorded; a real prompt/model *change* through the harness will follow the first such change.
- **#98:** shadow-mode exit criteria (≥10 consecutive clean scheduled runs, market-narrative gate pass rate ≥80%, human spot-review) and the go/no-go checkpoint remain objectively unmet and are unaffected by this PR.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_013BVhETBW4VsEehA3SYBKYh

---
_Generated by [Claude Code](https://claude.ai/code/session_013BVhETBW4VsEehA3SYBKYh)_

### CODE REVIEW

**Review by dceoy — COMMENTED** [review]
Reviewed PR #111 at `9404965748b5c6f4f7d817a9ac80e3dc5b97036b`.

Overall, the implementation direction is sound: stance evaluation stays deterministic, qualitative/prompt accountability is kept dependency-light, CI is green, and the OKF curation flow remains human-reviewed rather than auto-mutating canonical OKF files.

I recommend holding merge for the two existing unresolved P2 findings:

1. **Move stance evaluation after qualitative generation.** The daily workflow currently evaluates and renders `data/performance/${STEM}.json` / `content/evaluation/_index.md` before the later shadow qualitative step writes `data/qualitative/${STEM}.json`. With qualitative enabled, the same PR can commit a newly generated qualitative artifact while the evaluation artifact/page omit it, so rerunning the evaluator after merge changes `metadata.inputs.qualitative_dates`, pending counts, and evaluated counts. Move `Evaluate qualitative stances` to after the qualitative step and before report/Hugo validation/commit so the committed evaluation artifact is reproducible from the committed same-run inputs.

2. **Make `validate_performance.py` enforce `metadata.config.return_consistency_tolerance`.** The produced artifact includes this value and the schema treats it as required, but the hand-rolled validator currently only validates `metadata.config.horizons`. Because the workflow describes the output as schema-validated and this tolerance controls which return chains are skipped as broken, the validator should reject missing/non-numeric/invalid tolerance values.

Non-blocking follow-up: the prompt-regression record separates the current prompt SHA/model from `artifact_provenance`. That is acceptable for a baseline fixture harness, but it does not prove the recorded artifact was generated by the current prompt/model. Consider either requiring provenance equality for non-fixture runs, or explicitly marking fixture-baseline entries so the CI guard’s semantics remain clear when the first real prompt/model change is adopted.

File: .github/workflows/daily-market-analysis.yml [hunk 1]
```diff
@@ -204,6 +204,26 @@ jobs:
         run: |
           uv run python .agents/skills/market-analysis/scripts/validate_history.py \
             --input "data/history/${STEM}.json"
+      # Deterministic stance evaluation (#97): derives forward returns from
+      # committed analysis artifacts only, joins them with any accumulated
+      # qualitative stances (safe empty state before those exist), and
+      # refreshes the public evaluation page. Daily interval only — the
+      # forward-return chain is defined over daily bars.
+      - name: Evaluate qualitative stances
+        if: steps.symbols.outputs.symbols != '' && env.INTERVAL == 'd'
```
[the remark this answers is not in the document]
Comment (dceoy) [reply .github/workflows/daily-market-analysis.yml]: Good catch, fixed in 8d18595: moved "Evaluate qualitative stances" to run after "Run qualitative analysis (shadow mode)" so today's own qualitative artifact (if generated) is on disk before the evaluator runs.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/aims/validate_performance.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+        return ["'metadata' must be a JSON object"], []
+    errors = [
+        f"metadata missing required key: {key!r}"
+        for key in _REQUIRED_META
+        if key not in metadata
+    ]
+    horizons: list[int] = []
+    config = metadata.get("config")
+    if isinstance(config, dict):
+        raw = config.get("horizons")
```
[the remark this answers is not in the document]
Comment (dceoy) [reply src/aims/validate_performance.py:139]: Good catch, fixed in 8d18595: `validate_performance.py` now requires `metadata.config.return_consistency_tolerance` to be a non-negative number, matching the schema. Added tests for missing/negative/wrong-type/zero values.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/workflows/daily-market-analysis.yml [hunk 3]
```diff
@@ -252,6 +252,41 @@ jobs:
...
+      # qualitative stances (safe empty state before those exist), and
+      # refreshes the public evaluation page. Daily interval only — the
+      # forward-return chain is defined over daily bars. Runs after the
+      # qualitative-analysis step so that today's own qualitative artifact
+      # (if generated) is already on disk and included in this run's
+      # evaluation inputs instead of appearing only on the next scheduled
+      # run.
+      - name: Evaluate qualitative stances
+        if: steps.symbols.outputs.symbols != '' && env.INTERVAL == 'd'
+        env:
```
Comment (dceoy) [inline .github/workflows/daily-market-analysis.yml:271]: This step is not guarded by `steps.qualitative.outcome`, unlike report rendering and Slack. Because the qualitative step above is `continue-on-error`, a failed validation can leave `data/qualitative/${STEM}.json` on disk; `evaluate_stances.py` then loads all qualitative JSONs and can either fail the whole workflow or count an unvalidated same-day stance artifact. Please exclude the same-run qualitative artifact unless the qualitative step succeeded, or make the evaluator skip untrusted/unreadable qualitative inputs with warnings.

**Review by dceoy — COMMENTED** [review]
Review result: one blocking concern remains before merge.

CI is green, and the previous ordering/validator fixes look directionally correct, but the daily workflow now runs stance evaluation after the qualitative shadow step while the qualitative step is explicitly `continue-on-error`. If qualitative generation succeeds but validation fails, an invalid `data/qualitative/${STEM}.json` can remain on disk. Report rendering and Slack notification already guard against that with `QUALITATIVE_OUTCOME == "success"`, but stance evaluation does not. Since `evaluate_stances.py` loads every `data/qualitative/*.json` and propagates `JSONDecodeError`/`ValueError` as a non-zero exit, this can turn the documented fail-open qualitative shadow path into a hard workflow failure, or worse, include an unvalidated same-day qualitative artifact in performance metrics.

Please keep stance evaluation aligned with the fail-open trust boundary: either exclude the same-run qualitative artifact unless `steps.qualitative.outcome == 'success'`, pass an explicit vetted artifact list, or make the performance loader skip unvalidated/unreadable qualitative artifacts with warnings. Historical committed qualitative artifacts can still be evaluated normally.

**Review by dceoy — COMMENTED** [review]
Reviewed PR #111 at `8d18595d6b7bf5494423245aa78dae97b935921f`.

The previous two P2 findings appear fixed: stance evaluation now runs after qualitative generation, and `validate_performance.py` enforces `metadata.config.return_consistency_tolerance`. CI is green.

I recommend holding merge for one remaining blocker:

- **Guard stance evaluation against failed same-run qualitative output.** The qualitative step is `continue-on-error`, and its validation can fail after `data/qualitative/${STEM}.json` has been written. The newly moved `Evaluate qualitative stances` step then runs unconditionally for daily runs and `evaluate_stances.py` loads all qualitative JSON files from `data/qualitative`. That can either fail the whole workflow or, worse, include an unvalidated same-day qualitative artifact in the performance artifact and public evaluation page. This breaks the existing fail-open/shadow-mode contract already enforced by report rendering and Slack, which only trust the qualitative artifact when `steps.qualitative.outcome == 'success'`.

Suggested fix: before running the evaluator, exclude/delete `data/qualitative/${STEM}.json` unless the qualitative step outcome is `success`; alternatively, make the evaluator accept an explicit list of trusted qualitative artifacts or validate/skip untrusted same-run inputs with warnings.

No other blocking issues found in the latest diff.

**Comment by dceoy** [comment]
## Fail-open trust boundary for stance evaluation (57c6e94)

Addressed the gap identified during review: `evaluate_stances.py` was reading every file under `data/qualitative/*.json` unconditionally, including a possible same-run artifact for today's date. Since the "Run qualitative analysis (shadow mode)" step is `continue-on-error`, a file can land on disk even when generation or its own `validate_qualitative.py` call failed — report rendering and Slack notification already guard against this via `steps.qualitative.outcome == 'success'`, but stance evaluation didn't.

**How the trust boundary is enforced:**

1. `src/aims/performance.py` now has `_load_qualitative_artifacts(qualitative_dir, as_of, exclude_dates)`, which loads daily qualitative artifacts up to `as_of` but skips any date in `exclude_dates` **even if the file is present on disk**, recording a warning (`"<date>: excluded a same-run qualitative artifact because the qualitative-analysis step did not report success"`). Only caller-specified dates are ever excluded, so historical, already-validated artifacts are untouched.
2. `generate()`/`main()` expose this as a repeatable `--exclude-qualitative-date YYYY-MM-DD` CLI flag.
3. `build_artifact()` gained `extra_warnings` to merge these load-time warnings into the committed artifact's `warnings` list (sorted, deterministic).
4. The workflow's "Evaluate qualitative stances" step now passes `--exclude-qualitative-date "${DATE}"` whenever `steps.qualitative.outcome != 'success'` — the same condition already used downstream — so a failed/unvalidated same-run artifact is excluded from evaluation and its own next-day (or later) validated version is picked up normally once it exists.

**Preserved:** safe empty-state behavior (no qualitative artifacts → zero counts + warning, unchanged), fully deterministic output for a given set of committed inputs, no new runtime dependencies, and `validate_performance.py`'s validation of committed performance artifacts is untouched (still rejects malformed artifacts — verified by existing + new tests).

**Tests added** in `tests/test_performance.py`: same-run artifact excluded on step failure while historical artifacts stay in; same-run artifact included on step success (no flag passed); the loader's own date/interval filtering and warning generation in isolation; `build_artifact` warning merge/sort; and that artifacts written through this path still pass `validate_performance.validate_artifact`.

**Validation:** `ruff format .`, `ruff check .`, `pyright .` (0 errors), `pytest` — **939 passed, 100% branch coverage** maintained. Workflow YAML re-checked with `yamllint`, `actionlint`, and `zizmor --offline` — no findings.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by dceoy — COMMENTED** [review]
Reviewed the latest head `57c6e9461c0b60b1385a1327791df296fa07a06f`.

Result: no additional blocking issues found.

The previous blockers are addressed:

- Stance evaluation now runs after the qualitative shadow step and passes `--exclude-qualitative-date ${DATE}` unless `steps.qualitative.outcome == 'success'`, preserving the fail-open trust boundary for same-run qualitative output.
- `validate_performance.py` now enforces `metadata.config.return_consistency_tolerance`, matching the schema and the artifact provenance requirement.
- CI/CD is green on the latest head.

The remaining prompt-regression fixture/baseline semantics are worth documenting further as a future hardening item, but I do not consider them merge-blocking for this PR.
